Topics/Tech_Stacks/Metaprogramming.md

```diff
@@ -0,0 +1,142 @@
+        # the class with the data as the arguments.
+        cls = globals()[name]
+        return cls(**data)
+
+    # this is the method that handles the logic of the message. We need to implement
+    # this method in all our Message subclasses.
+    # for example class SignIn(username: str, password: str) would return a SignInResponse object.
+    # this is how we get dynamic dispatch to work. We can call the respond method on any Message object and it will return the correct response. as defined by the subclass.
+    def respond(self) -> Message:
+        raise NotImplementedError
+```
+
+Then assuming we have a server object that can send and receive JSON. 
+the server architecture is as follows.
+
+```python
+    while True:
+        json: str = server.get_json()
+        # we convert the json to a Message object using the from_json_message method.
+        message_as_cls: Message = Message.from_json_message(json)
+        # we use dynamic dispatch to call the respond method of the Message object.
+        response_as_cls: Message = message_as_cls.respond()
+        # we convert the response to json and send it back to the client.
+        response_json = response_cls.to_json_message()
+        server.send_json(response_json)
+```
+
+remember to import at the top of the file.
+
+```python
+from __future__ import annotations
+from datetime import datetime
+from typing import *
+import json
+
+from pydantic import BaseModel
+```
+
+Now all we need to do to add a new message type is to create a new class that inherits from Message and implement the respond method.
+
+```python
+
+# a simple example of the kind of message we might want to receive from the client.
+class SignIn(Message):
+    name: str
+    password: str
+
+    # this handles the logic of the message. This is where we would do things like check if the user exists in the database.
+    # or if the password is correct. and then we return a SignInResponse object.
+    def respond(self) -> SignInResponse:
+        return SignInResponse(success=True)
+
+
+# a simple example of the kind of message we might want to send to the client.
+# node we don't need to implement any logic here. We just need to define the fields we want to send.
+# we could implement validation logic through pydantic if we wanted too. All our messages are valid pydantic models.
+class SignInResponse(Message):
+    success: bool
+    valid_user: bool = True
+    valid_password: bool = True
+    error_message: Optional[str] = None
+```
+
+### why you want to do this.
+
+While there are many frameworks that do this for you. I think there are many advantages to doing it yourself.
+Addiotionally meta programming in python is incredibly powerful and can be used to do many things.
+Even if you won't use it for this specific purpose. I think it is a good thing to know about.
+### further things needed to cover.
+
+- how to generate a json schema.
+- how to automatically add items to a database.
```

Comment:
Consider adding a section on best practices or common pitfalls in metaprogramming for server architecture at the end. Also consider adding additional resources such as links to documentations or tutorials you used in writing your wiki post.

Reply:
I added a couple of further recourses on where the ideas in the wiki originated from. But a decent chunk of this architecture is my idea.